[clang][DebugInfo] Attach DW_AT_const_value to static data-member definitions if available

Michael137 opened the pull request:
In https://github.com/llvm/llvm-project/pull/71780 we started emitting definitions for all static data-members with constant initialisers, even if they were constants (i.e., didn't have a location). We also dropped the `DW_AT_const_value` from the declaration to help resolve [inconsistencies during type merging in the DWARFParallelLinker](https://github.com/llvm/llvm-project/pull/68721). However, for static data members that do have locations, we wouldn't emit a `DW_AT_const_value` on it, assuming that the consumer knows how to read the value using the location. This broke some consumers that really wanted to find a `DW_AT_const_value`.

This patch makes sure we also attach the `DW_AT_const_value` to definitions that have a location.

**Test changes**

* `debug-info-static-member.cpp`:
  * We added the check for `const_b` as part of the
    patch series in `638a8393615e911b729d5662096f60ef49f1c65e`.
    The check for `isUsableAsConstantExpression` added in the current patch
    doesn't support constant inline floats (since they are neither constexpr nor
    integrals). This isn't a regression since before said patch series
    we wouldn't ever emit the definition for `const_b` anyway. Now
    we just don't do it for `inline const float`s. This is consistent with
    GCC's behaviour starting with C++11.

* `debug-info-static-inline-member`:
  * This was just a bug which is now fixed. We shouldn't emit
    a `DW_AT_const_value` for a non-const static.

Michael137 committed:
[clang][DebugInfo][NFC] Create evaluateConstantInitializer helper function

Michael137 committed:
[clang][DebugInfo] Attach DW_AT_const_value to static data-member definitions if available

Michael137 committed:
fixup! clang-format

Michael137 committed:
fixup! make helper static

Michael137 commented:
@ilovepi

ilovepi commented:
This looks great, thanks for keeping me in the loop. :)

jryans commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -5503,11 +5516,17 @@ void CGDebugInfo::EmitGlobalVariable(llvm::GlobalVariable *Var,
+    if (Expr.empty()) {
+      if (auto const *InitVal = evaluateConstantInitializer(D))
+        E = createConstantValueExpression(D, *InitVal);
+    } else
```
I believe code style says this else block should use braces because the if block has them.

jryans reviewed (commented):
Overall this looks good, thanks!

I think we may want to add a test to check that the output DWARF has a const value in the expected place, since consumers are relying on it.

Michael137 commented:
> Overall this looks good, thanks!
> 
> I think we may want to add a test to check that the output DWARF has a const value in the expected place, since consumers are relying on it.

Any suggestions where to put such end-to-end test?

This patch makes sure we encode the constant in a `DIExpression` on definitions, which is tested. And we have tests for ensuring that `DW_OP_constu` expressions get turned into `DW_AT_const_value`s (I know of at least `llvm/test/DebugInfo/Generic/dwarf5-debug-info-static-member.ll`)

Michael137 committed:
fixup! add braces to adhere to llvm-code style

jryans reviewed (approved):
> This patch makes sure we encode the constant in a `DIExpression` on definitions, which is tested. And we have tests for ensuring that `DW_OP_constu` expressions get turned into `DW_AT_const_value`s (I know of at least `llvm/test/DebugInfo/Generic/dwarf5-debug-info-static-member.ll`)

Ah okay, then with that in mind, I think the test changes here are likely enough to merge. 🙂

Michael137 committed:
fixup! use getAnyInitializer instead of canonical decl

This is more idiomatic

Michael137 committed:
fixup! fix tests

* `debug-info-static-member.cpp`:
  * We added the check for `const_b` as part of the
    patch series in `638a8393615e911b729d5662096f60ef49f1c65e`.
    The new check `isUsableAsConstantExpression` doesn't support
    constant inline floats (since they are neither constexpr nor
    integrals). This isn't a regression since before said patch series
    we wouldn't ever emit the definition for `const_b` anyway. Now
    we just don't do it for `const float`s. This is consistent with
    GCC's behaviour starting with C++11.

* `debug-info-static-inline-member`:
  * This was just a bug which is now fixed. We shouldn't emit
    a `DW_AT_const_value` for a non-const static.

pogo59 commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -69,6 +69,29 @@ static uint32_t getDeclAlignIfRequired(const Decl *D, const ASTContext &Ctx) {
+/// Given a VarDecl corresponding to either the definition or
+/// declaration of a C++ static data member, if it has a constant
+/// initializer and is evaluatable, return the evaluated value.
+/// Returns std::nullopt on failure.
```
Maybe instead of "on failure" say "otherwise." It's not really a _failure_ condition.

pogo59 reviewed (approved):
One comment, otherwise LGTM.

Michael137 committed:
fixup! clarify function comment

Michael137 commented:
This will need a slight tweak in DwarfCompileUnit because it currently doesn't attach a DW_AT_location to global variable DIEs that already have a DW_AT_const_value

dwblaikie commented:
Hmm, I can't quite tell from the test case updates in the patch, at least at a glance: How does this get encoded at the IR level? Do we end up with two DIGlobalVariableExpressions? One with the constant value expresison, and one that references the actual global variable? Or somehow encode both in one?

Michael137 commented:
> Hmm, I can't quite tell from the test case updates in the patch, at least at a glance: How does this get encoded at the IR level? Do we end up with two DIGlobalVariableExpressions? One with the constant value expresison, and one that references the actual global variable? Or somehow encode both in one?

My understanding was that the DIExpression parameter to DIGlobalVariableExpression was empty for global variables with locations. So the patch just encodes the constant into that expression if it's otherwise empty.

My plan was to make [DwarfCompileUnit](https://llvm.org/doxygen/DwarfCompileUnit_8cpp_source.html#l00203) support global variables with locations *and* constants, which it currently assumes doesn't happen.

Let me know if I missed something here

dwblaikie commented:
> My understanding was that the DIExpression parameter to DIGlobalVariableExpression was empty for global variables with locations. So the patch just encodes the constant into that expression if it's otherwise empty.

I think in theory it can be non-empty (see what happens under a merge globals optimization... I'm not sure of an example of when this optimization fires, or what the debug info we emit looks like - in theory it should look like a global with a non-empty expression that describes offsetting the pointer forward to reach the global inside the merged global) & so then you'd have a hard time telling whether the expression is meant to be used in addition to the location, or as part of evaluating the location.

We don't really have a mechanism for encoding a variable in two locations (or a constant and a location) at the same time, I think. We could invent a special opcode to use in the expression to communicate this, or define some special handling if there's two separate expressions providing a location (or a location and a constant in this case) for the same variable (say that they're both valid, and emit them both if we can). 

@adrian-prantl thoughts?

dwblaikie commented:
https://github.com/llvm/llvm-project/blob/abd85cd473afedf112bf00630a22382fee4a7853/llvm/lib/CodeGen/GlobalMerge.cpp#L531 - this is around where I think we'd get a global with a location and a non-empty expression

Michael137 commented:
> > My understanding was that the DIExpression parameter to DIGlobalVariableExpression was empty for global variables with locations. So the patch just encodes the constant into that expression if it's otherwise empty.
> 
> I think in theory it can be non-empty (see what happens under a merge globals optimization... I'm not sure of an example of when this optimization fires, or what the debug info we emit looks like - in theory it should look like a global with a non-empty expression that describes offsetting the pointer forward to reach the global inside the merged global) & so then you'd have a hard time telling whether the expression is meant to be used in addition to the location, or as part of evaluating the location.
> 
> We don't really have a mechanism for encoding a variable in two locations (or a constant and a location) at the same time, I think. We could invent a special opcode to use in the expression to communicate this, or define some special handling if there's two separate expressions providing a location (or a location and a constant in this case) for the same variable (say that they're both valid, and emit them both if we can).
> 
> @adrian-prantl thoughts?

Thanks for the pointers. The closest I could get to triggering a GlobalMerge of const static data-members was for following code:
```
struct Foo {
    static const int a;
    static const int b;
};

const int Foo::a = 100;
const int Foo::b = 200;
```

But only if I mark the constants as `internal` in the IR as follows (I attached the whole file):
```
%struct.Foo = type { i8 }

@a = internal constant i32 100, align 4, !dbg !0
@b = internal constant i32 200, align 4, !dbg !5

define void @use1() {
  %x = load i32, ptr @a, align 4
  %y = load i32, ptr @b, align 4
  ret void
}
...
```

Then run `opt` as:
```
./bin/opt  -global-merge -global-merge-max-offset=100 -global-merge-on-const -S merge.ll
```

That changes the IR into:
```
...
@_MergedGlobals = private constant <{ i32, i32 }> <{ i32 100, i32 200 }>, align 4, !dbg !0, !dbg !13

@a = internal alias i32, ptr @_MergedGlobals
@b = internal alias i32, getelementptr inbounds (<{ i32, i32 }>, ptr @_MergedGlobals, i32 0, i32 1)
...
!1 = distinct !DIGlobalVariable(name: "a", linkageName: "_ZN3Foo1aE", scope: !2, file: !3, line: 6, type: !7, isLocal: false, isDefinition: true, declaration: !12)
...
!5 = !DIGlobalVariableExpression(var: !6, expr: !DIExpression())
!6 = distinct !DIGlobalVariable(name: "b", linkageName: "_ZN3Foo1bE", scope: !2, file: !3, line: 7, type: !7, isLocal: false, isDefinition: true, declaration: !9)
...
!13 = !DIGlobalVariableExpression(var: !6, expr: !DIExpression(DW_OP_plus_uconst, 4))
```

Michael137 commented:
> Do we end up with two DIGlobalVariableExpressions?

Just checked. With this patch, MergeGlobals would produce following IR:
```
...
!10 = !DIGlobalVariableExpression(var: !11, expr: !DIExpression(DW_OP_plus_uconst, 4, DW_OP_constu, 200, DW_OP_stack_value))
!11 = distinct !DIGlobalVariable(name: "b", linkageName: "_ZN12_GLOBAL__N_13Foo1bE", scope: !2, file: !3, line: 8, type: !4, isLocal: true, isDefinition: true, declaration: !9)
...
!14 = !DIGlobalVariableExpression(var: !11, expr: !DIExpression(DW_OP_constu, 200, DW_OP_stack_value))
...
```
I.e., two `DIGlobalVariableExpression`s for the same `DIGlobalVariable`, which AFAICT would confuse `DwarfDebug` at the moment, which expects there to be a 1-to-1 mapping.

dwblaikie commented:
> I.e., two `DIGlobalVariableExpression`s for the same `DIGlobalVariable`, which AFAICT would confuse `DwarfDebug` at the moment, which expects there to be a 1-to-1 mapping.

Would confuse `DwarfDebug` without your changes/this patch? Or only with this patch?
(& hopefully there's some way to reach the merge globals functionality/problem/interesting IR from clang itself - but I guess even if it isn't, this is still a demonstration of representational issues)

Michael137 commented:
> @Michael137 is this patch ready to land? We (I have took over the bug from @ilovepi ) have a few builders that are currently blocked by the behavior change introduced in PR #70639 . It would be great if this change can be landed timely. Thx.

Nope this isn't ready yet. We need to figure out how to make global variables with locations *and* constants work. I'm about to submit a separate PR that will unblock your builds

Michael137 commented:
> Would confuse `DwarfDebug` without your changes/this patch? Or only with this patch? (& hopefully there's some way to reach the merge globals functionality/problem/interesting IR from clang itself - but I guess even if it isn't, this is still a demonstration of representational issues)

Would only confuse `DwarfDebug` *with* this patch.
